Add minor and patch release information. (#227)
opensearch-project/.github: RELEASING.md

- [Overview](#overview)
- [Branching](#branching)

>   - [OpenSearch and OpenSearch Dashboards Branching](#opensearch-and-opensearch-dashboards-branching)

Reviewer: how about other repo

Author: right below.

  - [Plugin Branching](#plugin-branching)
  - [Single Repo Artifacts Branching](#single-repo-artifacts-branching)
  - [Feature Branches](#feature-branches)
  - [Backporting](#backporting)
- [Versioning](#versioning)
  - [Version Numbers](#version-numbers)

>     - [OpenSearch and OpenSearch Dashboards Version Numbers](#opensearch-and-opensearch-dashboards-version-numbers)

Reviewer: in our 2024 roadmap, we plan to decouple opensearch release and opensearch dashboards release. would suggest to check that and think through if we should define process to couple these two

Reviewer 2: This is as part current architecture. @seraphjiang Once we have proposal for decoupling accepted and executed for the project - we can revisit then.

  - [Other Projects](#other-projects)
  - [Tagging](#tagging)
  - [Release Labels](#release-labels)
  - [Changing Release Dates](#changing-release-dates)
  - [Requesting a Release](#requesting-a-release)
  - [Security Reviews](#security-reviews)

## Overview

This document explains the release strategy for artifacts in this organization, including large distributions, such as OpenSearch with OpenSearch Dashboards, and smaller distributions, such as clients.

## Branching

Projects create a new branch when they need to start working on 2 separate versions of the product, with the `main` branch being the furthermost release. 

### OpenSearch and OpenSearch Dashboards Branching

Most repos in this organization, including [OpenSearch](https://github.com/opensearch-project/OpenSearch) and [OpenSearch-Dashboards](https://github.com/opensearch-project/OpenSearch-Dashboards), typically track 3 releases in parallel. For example, given the last major release of 2.0, OpenSearch and all its components in this organization maintains the following active branches.

* **main**: The _next major_ release, currently 3.0. This is the branch where all merges take place, and code moves fast.

> * * **2.x**: The _next minor_ release. Once a change is merged into `main`, decide whether to backport it to `2.x`.

Reviewer: where is the branch for next patch release? could we add that, i would suggest to use 2.0.x

Reviewer 2: How will that be helpful?

Author: I added

```
* * **2.17**: The _last minor_ release and the _next patch_ release. Once a change is merged into `2.x`, it may be backported to, for example, `2.17` for the next 2.17 release candidate, or as a patch for a `2.12.1`.
```

Reviewer: @dblock 

* * **2.x**: The _next minor_ release
`2.17` for the next 2.17 release candidate, or as a patch for a `2.12.1`.

what's the different between _next minor_ releae compare to release candidate?

if 2.x if for next minor release, why could we add 2.17.x for next patch release, then we should use 2.17.1 for release candidate 

i know opensearch is using 2.17.1 for tag. but it is not very common compare to other opensource project.

Author: On 2.x you usually have the next minor release. Then eventually we cut 2.17 and make last minute fixes there to make the 2.17 release. So they are the same thing.

Then we make patches on top of 2.17 in the 2.17 branch while 2.x will have moved to 2.18. And release 2.17.1 from 2.17 branch.

Suggest a change if that's not clear from my text?

Frequently released artifacts, such as most [clients](https://github.com/opensearch-project/opensearch-clients), follow a simpler branching model with the next release always living on `main` and no patch branches.

### Feature Branches

Do not creating branches in the upstream repo, use your fork, for the exception of long lasting feature branches that require active collaboration from multiple developers. Name feature branches `feature/<thing>`. Once the work is merged to `main`, please make sure to delete the feature branch.

### Backporting

Backwards-incompatible changes always result in a new major version and will __never__ be backported. Small improvements and features will be backported to a new minor version (e.g. `2.9.0`). High severity security or critical bug fixes will be backported to a new patch version (e.g. `2.9.1`). Repos in this organization typically use a backport workflow where you can label a PR, e.g. `backport 2.x`, and the workflow will attempt an automatic backport and open a new PR. If the backport fails, it is the contributor's responsibility to do a manual backport following the instructions in the failed backport error message.

## Versioning

> All distributions in this organization [follow semver](https://opensearch.org/blog/technical-post/2021/08/what-is-semver/). A user-facing breaking change can only be made in a major release. Any regression that broke semver is considered a high severity bug. 

Reviewer 3: Any regression that breaks semver?  Also, [according to the spec](https://semver.org/) it's "SemVer"

Author: Fixed both.

> All distributions in this organization [follow SemVer](https://opensearch.org/blog/technical-post/2021/08/what-is-semver/). A user-facing breaking change can only be made in a major release. Any regression that breaks SemVer is considered a high severity bug.

Reviewer 4: I like taking a firm stance on this, also this isn't being followed in all repositories - opensearch-migration that I maintain isn't following semver.  

Do we have thoughts on how we roll out this **P**olicy change is known?

Author: AFAIK this has been the policy. I think this is a bug in opensearch-migrations, you should open it!

Reviewer 4: - https://github.com/opensearch-project/opensearch-migrations/issues/1042

The build number of the engine is 3-digit `major.minor.patch` (e.g. `2.9.0`), while plugins use 4 digits (`2.9.0.45`). See [OpenSearch#1093](https://github.com/opensearch-project/OpenSearch/issues/1093) for a proposal to remove this difference. 

#### Single Repo Version Numbers

The version number of all releasable artifacts is 3-digit `major.minor.patch` (e.g. `2.9.0`).

### Incrementing Versions

Versions are incremented as soon as development starts on a given version to avoid confusion. In the examples above versions are as follows.

* OpenSearch: `main` = 3.0.0, `2.x` = 2.10.0 `1.x` = 1.4.0, and `1.2` = 1.2.5
* job-scheduler: `main` = 3.0.0.0, `2.x` = 2.10.0.0 , `1.x` = 1.4.0.0 and `1.2` = 1.2.5.0

## Releases

### Major Releases

Major releases contain breaking changes, new features, and bug fixes. All artifacts released in this organization only do major releases when there are significant breaking changes ready for release that warrant a major upgrade. To avoid major upgrades we strive to build features in a backwards compatible way as much as possible, therefore it's common for a distribution to go multiple years without a major release.

> For example, the OpenSearch bundle that contains OpenSearch and OpenSearch Dashboards releases new major versions after Lucene major version releases. This is announced in the forums at least 4 weeks prior to the release date. Once we become aware of the need for a major version, we start a major version release window which is similar to a minor release, except for two things: 1) participation is mandatory for all components, and 2) the release window is at least 4 weeks long to accommodate the testing required.

Reviewer 3: So whaddya do if a component doesn't participate?  Drop it?   With some components down to a few maintainers, this is a going concern.

Author: The release team picks it up.

### Plugin Branching

> OpenSearch and OpenSearch Dashboards plugins are bundled and shipped together along with the OpenSearch distribution for every release. Plugin branching follows OpenSearch branching described above that allows working on 3 releases at the same time.

Reviewer: again, OSD release should decouple from opensearch release in the furture(2025), plugin release should decoupled with OpenSearch and OSD as well. at least OSD

Reviewer 2: Can you share the proposal here?

Author: we can update this when this happens

### Single Repo Artifacts Branching

Frequently released artifacts, such as most [clients](https://github.com/opensearch-project/opensearch-clients), follow a simpler branching model with the next release always living on `main` and no patch branches.

### Feature Branches

Do not creating branches in the upstream repo, use your fork, for the exception of long lasting feature branches that require active collaboration from multiple developers. Name feature branches `feature/<thing>`. Once the work is merged to `main`, please make sure to delete the feature branch.

### Backporting

> Backwards-incompatible changes always result in a new major version and will __never__ be backported. Small improvements and features will be backported to a new minor version (e.g. `2.9.0`). Security fixes will be backported to a new patch version (e.g. `2.9.1`). Repos in this organization typically use a backport workflow where you can label a PR, e.g. `backport 2.x`, and the workflow will attempt an automatic backport and open a new PR. If the backport fails, it is the contributor's responsibility to do a manual backport following the instructions in the failed backport error message.

Reviewer: missing critical bug for patch version release. could we add?
also do we mentioned critical for security and critical bug

Reviewer 2: OpenSearch follows [Semver](https://semver.org/) - ```PATCH version when you make backward compatible bug fixes```

Author: @seraphjiang I don't think I follow, can you please use the suggest button and add the text you think we should have here?

Reviewer: ```suggestion
Backwards-incompatible changes always result in a new major version and will __never__ be backported. Small improvements and features will be backported to a new minor version (e.g. `2.9.0`). Security or critical bug fixes will be backported to a new patch version (e.g. `2.9.1`). Repos in this organization typically use a backport workflow where you can label a PR, e.g. `backport 2.x`, and the workflow will attempt an automatic backport and open a new PR. If the backport fails, it is the contributor's responsibility to do a manual backport following the instructions in the failed backport error message.
```

## Versioning

All distributions in this organization [follow semver](https://opensearch.org/blog/technical-post/2021/08/what-is-semver/). A user-facing breaking change can only be made in a major release. Any regression that broke semver is considered a high severity bug. 

### Version Numbers

#### OpenSearch and OpenSearch Dashboards Version Numbers

The build number of the engine is 3-digit `major.minor.patch` (e.g. `2.9.0`), while plugins use 4 digits (`2.9.0.45`). See [OpenSearch#1093](https://github.com/opensearch-project/OpenSearch/issues/1093) for a proposal to remove this difference. 

#### Single Repo Version Numbers

> The build number of all releasable artifacts is 3-digit `major.minor.patch` (e.g. `2.9.0`).

Reviewer: let's clarify what is build number. typically it is to identify build, ideally we should have
2.9.0.(12345) as full version number.

12345 could be a build number generated by build automation and it is always increase.

Author: These are always 0 in OpenSearch, so I don't think this is correct.

Reviewer: if build name is not correct, should we use below 
```suggestion
The version number of all releasable artifacts is 3-digit `major.minor.patch` (e.g. `2.9.0`).
```

### Incrementing Versions

Versions are incremented as soon as development starts on a given version to avoid confusion. In the examples above versions are as follows.

* OpenSearch: `main` = 3.0.0, `2.x` = 2.10.0 `1.x` = 1.4.0, and `1.2` = 1.2.5
* job-scheduler: `main` = 3.0.0.0, `2.x` = 2.10.0.0 , `1.x` = 1.4.0.0 and `1.2` = 1.2.5.0

## Releases

### Major Releases

Major releases contain breaking changes, new features, and bug fixes. All artifacts released in this organization only do major releases when there are significant breaking changes ready for release that warrant a major upgrade. To avoid major upgrades we strive to build features in a backwards compatible way as much as possible, therefore it's common for a distribution to go multiple years without a major release.

> OpenSearch releases new major versions only when there are a critical mass of breaking changes (e.g. changes that are incompatible with existing APIs). These tend to be tied to Lucene major version releases, and will be announced in the forums at least 4 weeks prior to the release date. Once we become aware of the need for a major version, we will start a major version release window which will be similar to a minor release, except for two things: 1) participation is mandatory for all components, and 2) the release window will be at least 4 weeks long to accommodate the testing required. 

Reviewer: let's call out the limitation of this appraoch, right now the distribution release bundled OpenSearch, OpenSearch Dashboards and plugin. In the case OpenSearch doesn't has major change, OSD has break change, it is not clear if we should let OpenSearch Dashboards hold the critical feature(with break) to release and wait for OpenSearch. Especially the break change is caused by Security issue like CVE.

Author: I changed it to an example.

_For example, the OpenSearch bundle that contains OpenSearch and OpenSearch Dashboards releases new major versions after Lucene major version releases. This is announced in the forums at least 4 weeks prior to the release date. Once we become aware of the need for a major version, we start a major version release window which is similar to a minor release, except for two things: 1) participation is mandatory for all components, and 2) the release window is at least 4 weeks long to accommodate the testing required._

### Minor Releases

Minor releases for large distributions such as OpenSearch are scheduled on regular intervals, contain new features and bug fixes, and must not contain any breaking changes.

### Patch Releases

> Patch releases are reserved for high-severity CVEs, critical bug fixes (e.g. causes users to rollback after an upgrade), or significant regressions in non-experimental features (e.g. feature unusable), and must not contain any new features.

Reviewer 2: Critical bugs should be limited to existing features (not newly introduced in the current minor release), where users who are already utilizing the feature must rollback due to issues such as broken functionality, significant performance degradation, or unexpected failures rendering the feature unusable.

Reviewer 2: It's important to emphasize existing features (not newly introduced) so that feature/component owners understand that incomplete or poorly developed features should not rely on the possibility of a patch release for completion. Also, with new feature we are not breaking user expectation, so it doesn't demand patch release either.

Author: Can you please suggest some text that I can copy-paste?

Reviewer: how about this

```suggestion
Patch releases are reserved for addressing high-severity CVEs, critical bug fixes (e.g., issues causing users to roll back after an upgrade), or significant regressions in non-experimental features (e.g., feature rendered unusable). These releases must not introduce any new features.

Whether a fix qualifies as a critical bug should be proposed by the PR contributor or issue creator and triaged by the maintainer, with input from the community. The severity and impact of a bug in defining a critical issue can vary based on the type of repository. For example, a UX bug might be critical for end-user customers and could prevent them from using a feature, requiring domain expertise to assess. In contrast, backend repository bugs are generally more straightforward to triage.
```

Author: Taking first paragraph.

For the second I shortened a bit

_Whether a fix qualifies as a critical bug should be proposed by the PR contributor or issue creator, and triaged by the maintainers of the repo with input from the community. The severity and impact of a bug in defining criticality of an issue can vary. For example, back-end APIs are fairly straightforward to triage, while a UX bug may prevent users from using a feature and may require some domain expertise to assess._

> We follow [OpenSSF's best practices](https://bestpractices.coreinfrastructure.org/en/criteria/0?details=true&rationale=true#0.vulnerabilities_fixed_60_days) for patching publicly known vulnerabilities and we make sure that there are no unpatched vulnerabilities of medium or higher severity that have been publicly known for more than 60 days in our actively maintained versions.

Reviewer: For OSD 1.3.x is still Node 10 , its lifecycle Ended 3 years and 4 months ago. we need a better version strategy and lifecycle management for long term support. in the case. i would suggest to decouple OpenSearch Dashboards and OpenSearch release. and we could end support to OSD 1.3 because it is not secure, and we could not upgrade without break. 

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d0b47bdab4e5d5df4ea1e80b6fb928cb027824cf/package.json#L511

Author: Anything that this PR needs changing?

Reviewer: ```suggestion
We follow [OpenSSF's best practices](https://bestpractices.coreinfrastructure.org/en/criteria/0?details=true&rationale=true#0.vulnerabilities_fixed_60_days) for patching publicly known vulnerabilities and we make sure that there are no unpatched vulnerabilities of medium or higher severity that have been publicly known for more than 60 days in our actively maintained versions.

Please note: The runtime for OpenSearch Dashboards 1.3.x is based on Node.js 10, which has been end-of-life (EOL) for 3 years. Due to this, critical bug fixes and security patches are handled on a best-effort basis, as upgrading Node.js for 1.3.x would introduce significant breaking changes. This limitation arises from the need to release OpenSearch and OpenSearch Dashboards together.

We are aware of this issue and have a roadmap item to decouple the releases and versions for OpenSearch and OpenSearch Dashboards. This will allow us to upgrade or manage the versions of OpenSearch Dashboards independently in the future. More details can be found in the [OpenSearch Project Roadmap for 2024–2025](https://opensearch.org/blog/opensearch-project-roadmap-2024-2025/#roadmap-theme-3-ease-of-use) section: Serverless Dashboards and migration. 
```

Author: This is way too specific for this doc. What's the GitHub issue that captures this detail? 

I could change "we follow" to "Except for well-documented exceptions such as xyz, we follow ...".

Reviewer: I linked roadmap, a github issue might not necessary, if words is too much, we keep it short? there are two keypoints here, first is security patch is best effort for previous major version (1.x or what term we name it, please help). second it doesn't scale if we bump whole bundle distribution if only one component need to bump. decoupling project and make them release-able independently is roadmap.

## Releasing

### OpenSearch Releases

See [Releasing OpenSearch](https://github.com/opensearch-project/opensearch-build/blob/main/README.md#releasing-opensearch) for the standard operating procedures of making an OpenSearch release.

#### Release Process

> OpenSearch uses a "release-train" model. For minor version, that train leaves approximately every six weeks we release a new minor version which includes all the new features and fixes that are ready to go. Having a set release schedule makes sure OpenSearch is releasing in a predictable way and prevents a backlog of unreleased changes.

Reviewer: we did 6 minor version release in 2023, planed 6 minor in 2024, so it is about every 8 weeks right.

Author: @getsaurabh02 is this correct?

#### Release Schedule

At the beginning of every year, the project will publish on [opensearch.org](https://opensearch.org/releases.html) the "release windows start" dates for the year for all the large distributions, including OpenSearch. These dates will be spaced out ~6 weeks.

#### Release Window

On release window start date:

1. We generate the first candidate with all plug-ins/components that have met the entrance criteria. If a plug-in/component hasn't met all of the criteria, we'll version bump the last released version and release that.  Once the release window opens and the first candidate is generated, no additional features can be added, and we will not delay the start of a release window for any plug-in/component.

> 2. During the release window we will do final quality testing, documentation and performance testing.  Bug fixes can be added in during this time, but no new features will be added. 

Reviewer: what's the different between final quality test and test we enforce for each commit/PR

Author: Not sure. Suggest text?

Reviewer: ```suggestion
2. During the release window, we will conduct final quality testing, documentation updates, and performance testing. Bug fixes can be added during this time, but no new features will be included. We are not yet able to run release testing against every commit of each repository. During the release process, we will only enforce that each component addresses failed tests by either fixing them or skipping flaky tests. Neither the 2.x nor the 2.17 branches will meet the release quality bar without going through the full release process, including testing. In the event that any third party wants to fork and release from the 2.x or 2.17 branches, they will need to ensure they run their own quality tests at their own risk.
```

Author: I think this is confusing, at least in the release window section because it talks about state of a 2.x branch. I'll think about where to incorporate it, but for now not adding.

* Release ticket is cut, and there's a forum post announcing the start of the window.
* [Any necessary security reviews](##Security-Reviews) are complete. 

##### Exit Criteria to Close Release Window

* Performance tests are run, results are posted to the release ticket and there no unexpected regressions
* Documentation has been fully reviewed and signed off by the documentation community.
* All integration tests are passing
* Release blog is ready

### Other Projects

Look for a RELEASING.md file in each repo, e.g. [opensearch-java/RELEASING.md](https://github.com/opensearch-project/opensearch-java/blob/main/RELEASING.md). This is typically a [1-click release process](https://github.com/opensearch-project/opensearch-build/blob/main/ONBOARDING.md#onboarding-to-universal--1-click-release-process) in which a maintainer pushes a tag to the repo, followed by automation performing the actual release.

### Tagging

> Create tags after a release that match the version number, `major.minor.patch`, without a `v` prefix.

Reviewer: could we elaborate why we choose to tag without `v`, it is very confusion compare to branch. large oss project like linux, react use v- for tag

https://github.com/torvalds/linux/tags
https://github.com/facebook/react/tags

Author: Below it says

_For a discussion on whether to add a prefixing `v` to release tags, see [#35](https://github.com/opensearch-project/.github/issues/35)._

Reviewer: Thanks, @dblock. It doesn't explain why we chose not to add the "v-" at the beginning, but I understand it was a concern. there are effort to support, but no one work on it. and the world have gotten used to it now. 

Good to know, but we should try in new repository. Especially for those repository we took the ownership for maintenance from other open source org. like fork new repo.

The review process consists of building a threat model for the proposed change and optionally engaging a specialist to perform additional testing, such as a penetration testing. This process is done in parallel and in private within the project, during development, and usually takes 4-10 weeks. A repository maintainer will assess the scope of the new changes, initiate and manage a security review, provide public updates, and, if needed, communicate privately by email with the contributors. Please add a note in your pull request if you believe a security review is warranted. 

Please see [opensearch-project/.github#81](https://github.com/opensearch-project/.github/issues/81) for a discussion on improving this and other security-related processes.​

## Backporting

> This project follows [semantic versioning](https://semver.org/spec/v2.0.0.html). Backwards-incompatible changes always result in a new major version and will __never__ be backported. Small improvements and features will be backported to a new minor version (e.g. `2.9.0`). Security fixes will be backported to a new patch version (e.g. `2.9.1`).

Reviewer: i'm glad to see we remove this
